### DISCUSSION TITLE
Support for splitting `specs` section of environment's `spack.yaml` into multiple files

### DISCUSSION THREAD

**Discussion by iarspider** [spack/spack#33055] [Ideas]
As discussed during [last telcon](https://github.com/spack/spack/wiki/Telcon%3A-2022-10-05), I'd like to discuss the ability to build  `specs` section of an environment from several separate files, similar to how one [can already do](https://spack.readthedocs.io/en/latest/environments.html#included-configurations) for other sections. 

One possible syntax to do it could be the following:
```
spack:
  specs:
  - "$include build"
  - "$include generators"
  - root@some-special-version
```

where `$include X` entries will be replaced by the contents of `X.yaml`, located in the same directory as `spack.yaml`. So, if `build.yaml` and `generators.yaml` contain the following:

```
$ cat build.yaml
specs:
- automake@1.16.3
- autoconf@2.71
- m4@1.4.18~sigsegv
- gmake@4.3~nls
- cmake@3.21.4~ownlibs~ncurses

$ cat generators.yaml
specs:
- pythia8@8.306+hepmc+hepmc3+lhapdf+madgraph5amc
- madgraph5amc@2.7.3.py3 +cms~ninja~collier+golem~syscalc nb_core=4
- sherpa@2.2.12+analysis+fastjet+hepmc2+rivet+lhapdf+blackhat+python+ufo+openloops+mpi+cms~hepmc3
```

the environment file will be equivalent to the following:
```
spack:
  specs:
  - automake@1.16.3
  - autoconf@2.71
  - m4@1.4.18~sigsegv
  - gmake@4.3~nls
  - cmake@3.21.4~ownlibs~ncurses
  - pythia8@8.306+hepmc+hepmc3+lhapdf+madgraph5amc
  - madgraph5amc@2.7.3.py3 +cms~ninja~collier+golem~syscalc nb_core=4
  - sherpa@2.2.12+analysis+fastjet+hepmc2+rivet+lhapdf+blackhat+python+ufo+openloops+mpi+cms~hepmc3
  - root@some-special-version
```

If we follow this approach (i.e. essentially preprocessing `spack.yaml` before using it), we don't need to worry about possible conflicts - we let environment handle them based on `spack:concretizer:unify` value.

**Comment by shahzebsiddiqui** [comment]
what if we make this a keyword called `definitions.yaml` that is defined something like this

```
$ cat definitions.yaml
definitions:
  build: 
    - automake@1.16.3
    - autoconf@2.71
    - m4@1.4.18~sigsegv
    - gmake@4.3~nls 
    - cmake@3.21.4~ownlibs~ncurses
  generate: 
    - pythia8@8.306+hepmc+hepmc3+lhapdf+madgraph5amc
    - madgraph5amc@2.7.3.py3 +cms~ninja~collier+golem~syscalc nb_core=4
    - sherpa@2.2.12+analysis+fastjet+hepmc2+rivet+lhapdf+blackhat+python+ufo+openloops+mpi+cms~hepmc3
specs:
  - $build
  - $generate
```

Then in your spack.yaml you have something like

```
spack:
  include:
  - file:///<path/to/definitions.yaml> 
```

I can see this useful since i have two `spack.yaml` one used for `spack ci` that pushes to buildcache and another that is used for production
